### DISCUSSION TITLE
Allow property setter to have multiple types

### DISCUSSION THREAD

**Discussion by royberris** [dotnet/csharplang#4697] [Language Ideas]
Example in pseudo code:

```c#
internal string attachments;

public List<string> Attachments
{
  get {
    return attachments.Split(';').ToList();
  }

  set { // default setter
    attachments = string.Join(';', value);
  }

  set<string> { // of type string
    attachments = value;  // so value is of type string
  }
}

public void MyMethod()
{
  Attachments = new List<string>()
  {
    "a.png", "b.png", "c.png"
  };

  Attachments = "a.png;b.png;c.png";
}
```

**Why would this be useful?**

There are scenarios where we can only save data one way, but want to present it another way. Obviously the "right" approach would be to actually save the `List<string>` in a database. But some systems work with old databases where the choice was made to save attachments in a single string. When creating code in you data layer you do want to to be represented as a `List<string>`. 

In my example think about two projects, one with has the UI (maybe a MVC application) and the other project is the data layer. Internally we are referring to the field `attachments` but the outside world should interact with the `Attachments` property. 

**Are there any other solutions for this?**

We can make a method as a setter instead of using a property. Of course properties are introduced to remove get/set methods like we see in Java. And I do think this would be a good extension on properties.

**Is this necessary?**

This is not a requirement but I could see how this would be an interesting feature for C#. And it should not be implemented for a very small use case. This was one I could think of on the spot, but I'm sure more creative people know how to give this a better use case and maybe expand my idea.

-----

I have no experience with language design nor with the technical knowledge on what it takes to implement this . It was on my mind and I thought it would be a good place to share it.

**Comment by BreyerW** [comment]
There was an idea to introduce generic indexers and that could be extended to properties which i think would be better idea since you could use generic type to do the same thing and also potentially apply constraints

**Reply by royberris** [reply]
Could you provide a link on this?

**Reply by BreyerW** [reply]
Actually theres already proposal for generic properties [https://github.com/dotnet/csharplang/discussions/471](https://github.com/dotnet/csharplang/discussions/471)

**Comment by canton7** [comment]
IMO, properties should behave more or less like fields from the point of view of an outsider. If I set a property to some value, I expect to be able to read that property and get the same value out again<sup>1</sup>.

Your proposal breaks this.

-----------

<sup>1</sup> Barring things like a validation exception being thrown, etc.

**Reply by royberris** [reply]
Wouldn't you get the same value out again? Only the data structure is different.

One could say that a list `[ "a", "b", "c" ]` has the same value as `"a;b;c;"` with a different data structure.

**Reply by canton7** [reply]
You could say that, but I don't agree with you. It contains some of the same information perhaps, but the property isn't behaving like a field. For example, I can't easily check that what I got out is the same as what I put in.

**Reply by canton7** [reply]
Fundamentally, you have your data (`["a", "b", "c"]`) and a serialized representation of it (`"a;b;c"`). Those two roughly equivalent, but they are not the same. If one of the functions of your type is to deserialize the serialized representation of your data, there should be an explicit method to do this:

```cs
thing.Attachments = Thing.Deserialize("a;b;c");
// or
thing.DeserializeAttachments("a;b;c");
```

... or some such.

**Reply by theunrepentantgeek** [reply]
> Wouldn't you get the same value out again? Only the data structure is different.

The type is an essential part of what you get out of a property. 

For example, it matters very much whether a property containing `10` is typed as an **int**, **byte**, **long**, **double**, **float**, or **decimal**. 

I believe explicit `Set()` methods are the correct solution to this problem, specifically because they make it clear that conversion is happening.

**Comment by HaloFour** [comment]
You could create a wrapper struct for the property which has conversion operators:

```cs
public struct Attachments {
    private readonly string value;

    private Attachments(string value) => this.value = value;

    public static implicit operator string(Attachments attachments) => attachments.value;
    public static implicit operator Attachments(string value) => new Attachments(value);
    public static implicit operator Attachments(List<string> list) => new Attachments(string.Join(";", list);
}

```

You could also expose two versions of the property that interact with the same backing fields:

```cs
public List<string> Attachments { get; set; }

public string AttachmentsString {
    get => string.Join(";", Attachments);
    set => Attachments = value.Split(";").ToList();
}
```

Which might work out better if these convenience conversions are intended to be exposed in the model.

> But some systems work with old databases where the choice was made to save attachments in a single string.

This sounds like a concern that should be abstracted by a repository or other facade.  There's no reason to expose the fact that you're using a particular storage strategy for a given property, regardless of whether or not that is legacy.

**Comment by canton7** [comment]
There's another aspect, which is that, as a developer, if I see:

```cs
thing.Attachments = "a;b;c";
```

... then I come to the conclusion that `Attachments` is of type `string`, and if I access it, I'll get a `string`. If I'm hunting around for a way to get a `List<string>` of attachments, I'll start looking elsewhere.
